azure-pipelines.yml

```
@@ -305,34 +305,65 @@ stages:
               versionSpec: '$(python.version)'
             displayName: 'Use Python $(python.version)'
 
-          - bash: python -m pip install --upgrade pip==20.2.4
+          - bash: python -m pip install --upgrade pip==21.0.1
             displayName: 'Update pip'
 
           - script: |
               sqlcmd -U sa -P "ReallyStrongPwd1234%^&*" -Q "CREATE DATABASE test_ci;" -o create_db_output.txt
 
           - script: |
+              pip install --requirement requirements-dev-util.txt --requirement requirements-dev-test.txt --requirement requirements-dev-sqlalchemy.txt
               pip install --requirement requirements.txt
-              pip install --requirement requirements-dev-test.txt --requirement requirements-dev-sqlalchemy.txt --requirement requirements-dev-util.txt
               pip install .
             displayName: 'Install dependencies'
 
           - script: |
               pip install pytest pytest-cov pytest-azurepipelines
-              pytest --mssql --no-postgresql --no-spark --napoleon-docstrings --junitxml=junit/test-results.xml --cov=. --cov-report=xml --cov-report=html
+              pytest --mssql --no-postgresql --no-spark --napoleon-docstrings --junitxml=junit/test-results.xml --cov=. --cov-report=xml --cov-report=html --ignore=tests/cli --ignore=tests/integration/usage_statistics
+            displayName: 'pytest'
+
+  - stage: cli_integration
+    dependsOn: [scope_check, lint, usage_stats_integration, required, db_integration]
```

Could this maybe just be dependent on `[scope_check, lint]` so that it can run in parallel to the other stages? To speed up CI testing.
```suggestion
    dependsOn: [scope_check, lint]
```

---

@anthonyburdi This is a good question -- it was done on purpose.  Here is the bulleted reasoning:
* CLI tests are in the category of integration tests, because CLI functions utilize the lower-level (`DataContext`) API.
* To save time, integration tests do not exercise every backend as long as the tests for the underlying platform API perform all the required tests for backends, different versions of `Python`, `pandas`, etc.
* Therefore, to insure that the CLI integration tests run on top of a platform, fully tested on a broad coverage of tests, the above dependency is created.

Thanks!

---

@alexsherstinsky - All of that totally makes sense. Though we do only allow PRs to merge that have passed all of the required/comprehensive/backend/compatibility checks anyway. So if we run this integration stage in parallel then we only risk wasting the CI resources to run this integration if it would have been pre-empted by an earlier failure -- not that other items would be untested. I don't currently know the economics / parallel configuration of our CI pipeline, but faster builds for expensive developers at the expense of more cheap robots sounds like a good tradeoff to me.

---

This is a non-blocking comment though - I just wanted to raise a potential developer efficiency opportunity here 😄

---

🙇